# THỰC HÀNH LAB 01 - LÀM QUEN VỚI PYTORCH
**Sinh viên:** Nguyễn Trương Cao Sơn
**Trường:** Đại học Sài Gòn

### Bài 1. Tạo Tensor cơ bản
**Yêu cầu:** Cho biết kích thước, số chiều và kiểu dữ liệu của Tensor a.

In [1]:
import torch
a = torch.tensor(
    [[1, 2, 3], 
    [4, 5, 6]], 
    dtype=torch.int32
    )

print('Tensor a:\n', a)
print('Kích thước:', a.shape)
print('Số chiều:', a.ndim)
print('Kiểu dữ liệu:', a.dtype)

Tensor a:
 tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)
Kích thước: torch.Size([2, 3])
Số chiều: 2
Kiểu dữ liệu: torch.int32


**Trả lời:**
- Kích thước: `(2, 3)`
- Số chiều: `2`
- Kiểu dữ liệu: `torch.int32`

### Bài 2. Tạo các Tensor thường dùng
**Yêu cầu:**
- Nhận xét sự khác nhau giữa torch.rand() và torch.randn().
- Tạo thêm một Tensor 5x5 chứa toàn giá trị 10.

In [3]:
import torch

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)

print(a)
print(b)
print(c)
print(d)
print(e)

f = torch.full((5, 5), 10)
print('Tensor 5x5 toàn 10:\n', f)

tensor([[0.9844, 0.2035, 0.7023, 0.6920],
        [0.6656, 0.8871, 0.2722, 0.1908],
        [0.7654, 0.1156, 0.9600, 0.8118]])
tensor([[-0.7069, -0.6070,  0.3227, -2.1986],
        [-1.7727, -0.9186,  0.1243,  0.9807],
        [ 0.1606, -0.4656,  0.5010, -2.9611]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
Tensor 5x5 toàn 10:
 tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


**Trả lời:**
- `torch.rand()`: Sinh số ngẫu nhiên theo phân phối đều trong khoảng [0, 1).
- `torch.randn()`: Sinh số ngẫu nhiên theo phân phối chuẩn tắc (mean = 0, std = 1).

### Bài 3. Kiểm tra thông tin Tensor
**Yêu cầu:** Giải thích ý nghĩa của torch.Size([2, 3, 4]).

In [4]:
import torch
a = torch.zeros(2, 3, 4)
print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


**Trả lời:**
- `torch.Size([2, 3, 4])` nghĩa là tensor 3 chiều. Chiều 0 có 2 phần tử, chiều 1 có 3 phần tử, chiều 2 có 4 phần tử. (Tổng có 2x3x4 = 24 phần tử).

### Bài 4. Slicing và Indexing
**Yêu cầu:**
- Cho biết shape của a[1] và a[1:, 2:4].
- Thử lấy hai phần tử cuối cùng trên chiều thứ ba.

In [6]:
import torch
a = torch.randn(3, 4, 5)
print("Original tensor:")
print(a)
print("\nTensor a[1]:")
print(a[1])
print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])

phan_tu_cuoi = a[:, :, -2:]
print("\nHai phần tử cuối cùng trên chiều thứ ba:\n", phan_tu_cuoi)
print("Shape của hai phần tử cuối:", phan_tu_cuoi.shape)

Original tensor:
tensor([[[-0.1011, -0.5730,  0.4330,  0.6479,  0.7273],
         [-0.5315,  0.5131, -1.0418,  0.5628,  0.5342],
         [-1.7856,  0.0052,  0.3024,  0.3201, -2.8984],
         [-0.0343,  1.9460,  0.2864,  1.2175, -0.9606]],

        [[ 1.3040,  0.8783, -0.6282, -1.3843, -0.1482],
         [ 0.9053, -1.1175,  0.6982, -0.4348,  1.0785],
         [ 0.4107,  0.7262,  1.4194,  1.0854, -0.1706],
         [ 0.2742,  2.1336,  0.2313, -0.9929, -0.5094]],

        [[-0.1589, -0.6907,  1.5446,  1.1176, -0.3674],
         [-0.4789, -1.1221,  0.7844,  0.7632, -0.0666],
         [-0.5603, -1.1454, -0.3305,  0.1982,  1.1709],
         [-0.5758, -0.7606,  1.3626,  0.4511,  0.8997]]])

Tensor a[1]:
tensor([[ 1.3040,  0.8783, -0.6282, -1.3843, -0.1482],
        [ 0.9053, -1.1175,  0.6982, -0.4348,  1.0785],
        [ 0.4107,  0.7262,  1.4194,  1.0854, -0.1706],
        [ 0.2742,  2.1336,  0.2313, -0.9929, -0.5094]])

Tensor a[1:, 2:4]:
tensor([[[ 0.4107,  0.7262,  1.4194,  1.0854, -0.1

**Trả lời:**
- Shape của `a[1]` là `torch.Size([4, 5])`.
- Shape của `a[1:, 2:4]` là `torch.Size([2, 2, 5])`.
- Code lấy 2 phần tử cuối: `a[:, :, -2:]`.

### Bài 5. Reshape và Flatten
**Yêu cầu:**
- Kiểm tra số phần tử trước và sau khi reshape.
- Thử reshape Tensor thành kích thước 2x6.

In [ ]:
import torch
a = torch.randn(3, 4)

print("Original:")
print(a)
print("\nFlatten:")
print(a.ravel())
print("\nReshape to 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

print("\nSố phần tử trước khi reshape:", a.numel())
print("Số phần tử sau khi reshape:", a.reshape(3, 2, 2).numel())

print("\nReshape thành kích thước 2x6:\n", a.reshape(2, 6))

Original:
tensor([[-0.8865, -1.1929, -0.0925, -1.0800],
        [-0.1380, -0.2881, -0.3158,  0.1086],
        [ 1.2573, -0.0222,  0.4832, -0.3377]])

Flatten:
tensor([-0.8865, -1.1929, -0.0925, -1.0800, -0.1380, -0.2881, -0.3158,  0.1086,
         1.2573, -0.0222,  0.4832, -0.3377])

Reshape to 3 x 2 x 2:
tensor([[[-0.8865, -1.1929],
         [-0.0925, -1.0800]],

        [[-0.1380, -0.2881],
         [-0.3158,  0.1086]],

        [[ 1.2573, -0.0222],
         [ 0.4832, -0.3377]]])

Số phần tử trước khi reshape: 12
Số phần tử sau khi reshape: 12

Reshape thành kích thước 2x6:
 tensor([[-0.8865, -1.1929, -0.0925, -1.0800, -0.1380, -0.2881],
        [-0.3158,  0.1086,  1.2573, -0.0222,  0.4832, -0.3377]])


**Trả lời Bài 5:**
- Số phần tử trước và sau khi reshape không thay đổi (đều bằng 12 phần tử).

### Bài 6. Phép toán và nhân ma trận
**Yêu cầu:**
- Phân biệt phép toán a * b với a @ b.T.
- Cho biết shape của kết quả a @ b.T.

In [8]:
import torch
a = torch.randn(2, 3)
b = torch.randn(2, 3)

print("a:")
print(a)
print("b:")
print(b)
print("\na + b:")
print(a + b)
print("\na / b:")
print(a / b)
print("\na^2:")
print(a ** 2)
print("\nMatrix multiplication:")
print(a @ b.T)

print("\nShape của kết quả a @ b.T:", (a @ b.T).shape)

a:
tensor([[ 0.2001,  0.3290,  0.4961],
        [ 0.4289, -0.3088,  0.0153]])
b:
tensor([[-1.5852, -0.9351,  0.9125],
        [ 1.5998, -1.3420, -0.0164]])

a + b:
tensor([[-1.3851e+00, -6.0604e-01,  1.4086e+00],
        [ 2.0288e+00, -1.6508e+00, -1.0569e-03]])

a / b:
tensor([[-0.1262, -0.3519,  0.5437],
        [ 0.2681,  0.2301, -0.9355]])

a^2:
tensor([[4.0037e-02, 1.0825e-01, 2.4613e-01],
        [1.8398e-01, 9.5344e-02, 2.3491e-04]])

Matrix multiplication:
tensor([[-0.1721, -0.1295],
        [-0.3772,  1.1004]])

Shape của kết quả a @ b.T: torch.Size([2, 2])


**Trả lời Bài 6:**
- `a * b` là phép nhân element-wise (nhân từng phần tử tương ứng của 2 ma trận với nhau).
- `a @ b.T` là phép nhân ma trận (matrix multiplication) theo đại số tuyến tính (dòng nhân cột).
- Shape của kết quả `a @ b.T` là `[2, 2]` do ma trận a (2x3) nhân với ma trận b.T (3x2).

### Bài 7. Các hàm thống kê trên Tensor
**Yêu cầu:** Thay dim=0 bằng dim=1 và nhận xét sự khác nhau.

In [10]:
import torch
a = torch.randn(3, 4)
print(a)

print("\nMean:")
print(torch.mean(a, dim=0))
print("\nStandard deviation:")
print(torch.std(a, dim=0))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=0))

print("\n--- Khi thay bằng dim=1 ---")
print("\nMean (dim=1):\n", torch.mean(a, dim=1))
print("\nStandard deviation (dim=1):\n", torch.std(a, dim=1))
print("\nCumulative sum (dim=1):\n", torch.cumsum(a, dim=1))

tensor([[-0.0101, -0.8182,  1.8538, -0.9452],
        [-0.1721,  0.1009,  0.2685, -0.3664],
        [-0.1676,  0.4245,  0.3443, -0.7696]])

Mean:
tensor([-0.1166, -0.0976,  0.8222, -0.6937])

Standard deviation:
tensor([0.0922, 0.6447, 0.8942, 0.2968])

Cumulative sum:
tensor([[-0.0101, -0.8182,  1.8538, -0.9452],
        [-0.1822, -0.7172,  2.1223, -1.3116],
        [-0.3498, -0.2927,  2.4666, -2.0812]])

--- Khi thay bằng dim=1 ---

Mean (dim=1):
 tensor([ 0.0201, -0.0422, -0.0421])

Standard deviation (dim=1):
 tensor([1.2907, 0.2822, 0.5514])

Cumulative sum (dim=1):
 tensor([[-0.0101, -0.8283,  1.0255,  0.0803],
        [-0.1721, -0.0711,  0.1974, -0.1690],
        [-0.1676,  0.2569,  0.6012, -0.1684]])


**Trả lời Bài 7:**
- Khi `dim=0`, các hàm thống kê được tính dọc theo các cột (từ trên xuống dưới của mỗi cột).
- Khi thay bằng `dim=1`, các hàm thống kê được tính dọc theo các hàng (từ trái qua phải của mỗi hàng).

### Bài 8. Autograd cơ bản
**Yêu cầu:**
- Tính thủ công đạo hàm của y = x^2 tại x = 3.6 và so sánh với x.grad.
- Đổi hàm thành y = 3*x**2 + 2*x + 1 và kiểm tra gradient.

In [11]:
import torch

x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()

print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

x_new = torch.tensor(3.6, requires_grad=True)
y_new = 3 * x_new**2 + 2 * x_new + 1
y_new.backward()

print("\n--- Với hàm y = 3*x**2 + 2*x + 1 ---")
print("x =", x_new)
print("y =", y_new)
print("x.grad =", x_new.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)

--- Với hàm y = 3*x**2 + 2*x + 1 ---
x = tensor(3.6000, requires_grad=True)
y = tensor(47.0800, grad_fn=<AddBackward0>)
x.grad = tensor(23.6000)


**Trả lời Bài 8:**
- Tính thủ công hàm $y = x^2$: $y' = 2x$. Tại $x = 3.6 \Rightarrow y' = 2 \times 3.6 = 7.2$. Kết quả khớp hoàn toàn với `x.grad`.
- Tính thủ công hàm $y = 3x^2 + 2x + 1$: $y' = 6x + 2$. Tại $x = 3.6 \Rightarrow y' = 6 \times 3.6 + 2 = 23.6$. Kết quả in ra của `x_new.grad` bằng 23.6, hoàn toàn chính xác.

### Bài 9. Gradient Descent tìm hệ số đa thức
**Yêu cầu:**
- So sánh ba hệ số học được với [1, 2, 3].
- Giải thích vai trò của optimizer.zero_grad(), mse.backward() và optimizer.step().

In [12]:
import numpy as np
import torch

polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([X * X, X, np.ones_like(X)])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w)

for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("\nLearned coefficients:")
print(w)

Initial coefficients:
tensor([[-0.0122],
        [ 0.1825],
        [ 1.0283]], requires_grad=True)

Learned coefficients:
tensor([[1.0072],
        [2.0118],
        [2.5328]], requires_grad=True)


**Trả lời Bài 9:**
- Ba hệ số học được đã hội tụ sát về giá trị [1, 2, 3] tương ứng với các hệ số của đa thức.
- `optimizer.zero_grad()`: Reset toàn bộ gradient của bước cập nhật trước đó về 0 để tránh bị cộng dồn.
- `mse.backward()`: Kích hoạt autograd, tính toán đạo hàm (gradient) của hàm loss (mse) theo từng trọng số (w).
- `optimizer.step()`: Cập nhật các trọng số theo hướng ngược lại của gradient để cực tiểu hóa sai số.

### Bài 10. Giải hệ phương trình bằng Autograd
**Yêu cầu:**
- Kiểm tra nghiệm thu được với bốn phương trình: A+B=9, C-D=1, A+C=8, B-D=2.
- Nhận xét vì sao bài toán có thể có nhiều nghiệm.

In [13]:
import random
import torch

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)

EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    
    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)

print("\n--- Kiểm tra nghiệm ---")
print(f"A + B = {A.item() + B.item():.4f} (Kỳ vọng: 9)")
print(f"C - D = {C.item() - D.item():.4f} (Kỳ vọng: 1)")
print(f"A + C = {A.item() + C.item():.4f} (Kỳ vọng: 8)")
print(f"B - D = {B.item() - D.item():.4f} (Kỳ vọng: 2)")

A = tensor(4.2561, requires_grad=True)
B = tensor(4.7438, requires_grad=True)
C = tensor(3.7439, requires_grad=True)
D = tensor(2.7438, requires_grad=True)

--- Kiểm tra nghiệm ---
A + B = 8.9999 (Kỳ vọng: 9)
C - D = 1.0000 (Kỳ vọng: 1)
A + C = 8.0000 (Kỳ vọng: 8)
B - D = 2.0000 (Kỳ vọng: 2)


**Trả lời Bài 10:**
- Bài toán có vô số nghiệm vì trong hệ 4 phương trình này có sự phụ thuộc tuyến tính lẫn nhau. Cụ thể: (A + B) - (B - D) = A + D = 7, và (A + C) - (C - D) = A + D = 7. Nên thực chất hệ chỉ tương đương với 3 phương trình độc lập cho 4 ẩn.

### Bài 11. Bài tập áp dụng
**Yêu cầu:**
- Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4x6, 2x12 và 2x3x4.
- Tạo hai Tensor A, B kích thước 3x4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột.
- Cho y = 2x^2 + 5x + 3. Dùng autograd tính dy/dx tại x=2 và kiểm tra bằng phép tính thủ công.
- Thay đa thức trong Bài 9 bằng y = 2x^2 - 3x + 5; huấn luyện và cho biết các hệ số tìm được.

In [14]:
import torch
import numpy as np

# Yêu cầu 1
X = torch.arange(1, 25)
print("Tensor X (4x6):\n", X.reshape(4, 6))
print("\nTensor X (2x12):\n", X.reshape(2, 12))
print("\nTensor X (2x3x4):\n", X.reshape(2, 3, 4))

# Yêu cầu 2
A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("\nA + B =\n", A + B)
print("\nA * B =\n", A * B)
print("\nA / B =\n", A / B)
print("\nTrung bình theo cột:\n", torch.mean(A, dim=0))
print("\nĐộ lệch chuẩn theo cột:\n", torch.std(A, dim=0))

# Yêu cầu 3
x = torch.tensor(2.0, requires_grad=True)
y = 2*x**2 + 5*x + 3
y.backward()
print("\ndy/dx tại x=2:", x.grad.item())
print("Tính thủ công: y' = 4x + 5. Tại x=2, y' = 4*2 + 5 = 13. (Khớp)")

# Yêu cầu 4
poly = np.poly1d([2, -3, 5])
X_dt = np.random.randn(20, 1) * 5
Y_dt = poly(X_dt)
XX_dt = np.hstack([X_dt * X_dt, X_dt, np.ones_like(X_dt)])

w_new = torch.randn(3, 1, requires_grad=True)
x_new = torch.tensor(XX_dt, dtype=torch.float32)
y_new = torch.tensor(Y_dt, dtype=torch.float32)

opt_new = torch.optim.NAdam([w_new], lr=0.1)

for _ in range(2000):
    opt_new.zero_grad()
    pred = x_new @ w_new
    loss = torch.mean(torch.square(y_new - pred))
    loss.backward()
    opt_new.step()

print("\nHệ số học được cho y = 2x^2 - 3x + 5:\n", w_new.data)

Tensor X (4x6):
 tensor([[ 1,  2,  3,  4,  5,  6],
        [ 7,  8,  9, 10, 11, 12],
        [13, 14, 15, 16, 17, 18],
        [19, 20, 21, 22, 23, 24]])

Tensor X (2x12):
 tensor([[ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12],
        [13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24]])

Tensor X (2x3x4):
 tensor([[[ 1,  2,  3,  4],
         [ 5,  6,  7,  8],
         [ 9, 10, 11, 12]],

        [[13, 14, 15, 16],
         [17, 18, 19, 20],
         [21, 22, 23, 24]]])

A + B =
 tensor([[-1.9346, -1.8990,  0.1554,  0.0376],
        [-3.3004, -4.0448,  0.9693,  1.3112],
        [-0.5015, -1.1126,  1.6122, -0.2459]])

A * B =
 tensor([[ 2.8103e-02, -1.4705e+00, -1.0653e+00, -3.8626e-04],
        [ 2.4032e+00,  4.0064e+00, -3.7826e-01,  4.2510e-01],
        [-2.0628e-02, -5.1581e-01,  5.7860e-01, -3.5437e-01]])

A / B =
 tensor([[ 7.6235e-03, -2.3725e-01, -1.1624e+00, -1.8266e-01],
        [ 2.0432e+00,  1.3337e+00, -2.3538e-01,  1.2340e+00],
        [-1.4121e+01, -2.4042e-01,  5.0262e-